File: LeetCode/442. Find All Duplicates in an Array.java
class Solution {
  public List<Integer> findDuplicates(int[] nums) {
    for (int i = 0; i < nums.length; i++) {
      int n = nums[i];
      while (n > 0) {
        nums[i] = 0;
        int tmp = nums[n - 1];
        if (nums[n - 1] > 0) {
          nums[n - 1] = -1;
        } else {
          nums[n - 1]--;
        }
        n = tmp;
      }
    }

    List<Integer> ans = new ArrayList<>();
    for (int i = 0; i < nums.length; i++) {
      if (nums[i] == -2) {
        ans.add(i + 1);
      }
    }

    return ans;
  }
}

File: LeetCode/442. Find All Duplicates in an Array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        for i, n in enumerate(nums):
            while n > 0:
                nums[i] = 0
                tmp = nums[n - 1]
                if nums[n - 1] > 0:
                    nums[n - 1] = -1
                else:
                    nums[n - 1] -= 1
                n = tmp

        ans = [i + 1 for i, n in enumerate(nums) if n == -2]
        return ans
